packages/api/src/services/emp-stats.ts

```diff
@@ -15,7 +18,42 @@ export default (config: Config, appState: Dependencies) => {
   const { currency = "usd" } = config;
 
   const queries = Queries(appState);
-  async function updateTvl(address: string) {
+
+  async function getLatestStats(address: string) {
+    return stats[currency].latest.getOrCreate(address);
+  }
+  function getOrCreateHistoryTable(address: string) {
+    let table = stats[currency].history[address];
+    if (table) return table;
+    table = empStatsHistory.SortedJsMap("Emp Stat History for " + address);
+    stats[currency].history[address] = table;
+    return table;
+  }
+  async function hasStatsHistory(address: string, timestamp: number) {
+    return getOrCreateHistoryTable(address).has(timestamp);
+  }
+  async function createStatHistory(address: string, data: EmpHistoryStat) {
+    return getOrCreateHistoryTable(address).create(data);
+  }
+
+  async function updateHistory(address: string) {
+    const stat = await getLatestStats(address);
+    assert(uma.utils.exists(stat.timestamp), "stats require timestamp");
```

Comment:
a question for my understanding: say one of these `asserts` fails. how does this manifest to the caller of the API? do they get a `500` internal server error with this assert message? where is this logic processed?

Reply:
so theres 2 categories of errors here
1. most services are responsible for ingestion, so errors in this particular service will just be logged to console.error and are not tied to client requests.  I am using the same pattern in pretty much every ingestion service for logging errors, [here](https://github.com/daywiss/protocol/blob/feature/ch1301/ability-to-obtain-a-time-series-of-tvl-of/packages/api/src/services/emp-stats.ts#L83)
2. the actions/express service are responsible for responding to client requests, any errors thrown here get sent back to the user with a status 500 and the error message.  You can see [here where express calls the actions,](https://github.com/daywiss/protocol/blob/feature/ch1301/ability-to-obtain-a-time-series-of-tvl-of/packages/api/src/services/express.ts#L32) catches it and passes it through.  Eventually its handled by [express error handler](https://github.com/daywiss/protocol/blob/feature/ch1301/ability-to-obtain-a-time-series-of-tvl-of/packages/api/src/services/express.ts#L44)

Reply:
makes sence! at some point replacing the `console.log` with an emit handler might be useful so we can capture the logs in GCP stackdriver